### PR TITLE
Provides a color picker in the UI to select desired Path color

### PR DESCRIPTION
**Pull request by willvincent**

### CODE REVIEW

**Comment by cbscribe** [comment]
This is very useful. The default blue color is very hard to see against some backgrounds.

**Comment by willvincent** [comment]
Fixes Issue #18164

**Comment by willvincent** [comment]
I suppose there's no reason the methods need to be named set/get_frame_color, I'm not intimately familiar with the codebase. I used ColorRect as a guide and it now occurs to me those method names may be more relevant for a rectangle.  

I'll change to set/get_color or something else if that makes more sense, just let me know :)

File: scene/2d/path_2d.cpp [hunk 1]
```diff
@@ -140,11 +148,15 @@ void Path2D::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("set_curve", "curve"), &Path2D::set_curve);
 	ClassDB::bind_method(D_METHOD("get_curve"), &Path2D::get_curve);
 	ClassDB::bind_method(D_METHOD("_curve_changed"), &Path2D::_curve_changed);
+	ClassDB::bind_method(D_METHOD("set_frame_color", "color"), &Path2D::set_frame_color);
+	ClassDB::bind_method(D_METHOD("get_frame_color"), &Path2D::get_frame_color);
 
+	ADD_PROPERTY(PropertyInfo(Variant::COLOR, "pathColor"), "set_frame_color", "get_frame_color");
```
Comment (Noshyaar) [inline scene/2d/path_2d.cpp]: the property should be `path_color` and as you mentioned the setter and getter name should be set/get_path_color

**Review by ghost — COMMENTED** [review]
- PR should be made against master branch
- self_modulate can do this if the default line color is white

**Comment by willvincent** [comment]
Ok, PR against correct branch, fixed method names.

Not sure I fully understand self_modulate, but I get the impression that will just automatically adjust the color. I don't know if that's necessarily better. I think there's some real value to being able to define specific path colors.

For example, if you had multiple paths for different purposes, you might want each of the same 'type' to be one color, and each of another type to be another color.

Also, I've no idea how self_modulate would be implemented here :)

**Review by volzhs — APPROVED** [review]
it looks good to me

**Comment by akien-mga** [comment]
Could you [squash the commits](http://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#mastering-the-pr-workflow-the-rebase) together?

**Comment by willvincent** [comment]
@akien-mga squashed as requested.

**Comment by groud** [comment]
That's a good feature, but I don't think it's a good idea to expose a function that is supposed to be an editor-only feature (the path color is not useful in games)

I would move out the color picking to a menu in the top bar for know (until we rewrite the inspector) and use the node's metadata to store the color.

**Comment by willvincent** [comment]
@groud that seems a lot more complicated than this. But I'm new to the codebase, and really have no idea how one would implement this as you suggest.

**Comment by groud** [comment]
> @groud that seems a lot more complicated than this. But I'm new to the codebase, and really have no idea how one would implement this as you suggest.

The problem is that if we start adding editor-only properties directly in the user API, this will needs us to maintain it. For editor-only things, we can remove/add metadata whenever we want, that's simpler to maintain.

To handle that, you probably have to add a menu entry in the top bar. This should be done in the path_2d_editor_plugin.cpp file. The color picker would then set a metadata on the node using `mySelectedPath2D.set_metatdata("_edit_color", selectedColor)`. The display code would just use this color if the metadata is present:
```
if(has_metadata("_edit_color"))
   color = get_metadata("_edit_color")
```

**Comment by willvincent** [comment]
I guess I don't understand how burying this setting in a menu is better to displaying it within the node's inspector, where it is contextually relevant to the active node.

Maybe I just don't understand your description, a diagram might help because frankly what you're suggesting sounds like terrible UX.

**Comment by groud** [comment]
But the problem is quite simple: we don't expose editor-only properties to the user, as otherwise it needs to be maintained. This is a priority to keep things simple in the long run.

That being said, yes, I agree it's a terrible UX. The same problem exists for a lot of other nodes (like the layout menus for control ones, that would better fit in the inspector). A rework of the inspector is planned, but until now, we have to use the top toolbar. Since keeping the API clean is a lot more important than UX (The API can't be changed over time while the editor can).

**Comment by willvincent** [comment]
You seem to be the only person that has a problem with this.

I think it's useful, several other people do as well. I'm leaving it here and walking away, merge it or don't. If the inspector/etc is being reworked in the future -- this can be changed then as well, there's really nothing to "maintain" here.

**Comment by groud** [comment]
> this can be changed then as well, there's really nothing to "maintain" here.

No it cannot, we cannot remove functions from the API for 3.x projects wince it might break compatibility.

**Comment by willvincent** [comment]
It occurs to me, your entire argument is based on the idea that path color doesn't matter in-game. While that it likely true most of the time, who's to say that is _always_ the case. There could be cases where displaying a path in-game would be useful, and controlling it's color would be even more useful in those cases.

Off the top of my head, suppose you have a mechanic where the player is able to see the defined path of a sentry, and maybe you want the color of that to pulse, or change depending on some current status...

**Comment by groud** [comment]
> There could be cases where displaying a path in-game would be useful, and controlling it's color would be even more useful in those cases.

Obviously, but that's not the point of core node. The core nodes are meant to cover the common case, not all of them. If it's not used often at all, we don't add it. Else the API would become bloated and a pain to maintain.

Regarding your use case, just follow the [new feature workflow](https://twitter.com/reduzio/status/941799197774315525). In that case, it's not a common use case, and can be workarounded by just exporting a color variable in a simple script. So there's no reason to add it as a built-in property (at least exposed to the user).

**Comment by willvincent** [comment]
Whatever.

**Comment by volzhs** [comment]
> The core nodes are meant to cover the common case, not all of them.

I agree with that.
I think it's better to move it to editor side in somehow.

**Comment by vnen** [comment]
> It occurs to me, your entire argument is based on the idea that path color doesn't matter in-game. While that it likely true most of the time, who's to say that is always the case. There could be cases where displaying a path in-game would be useful, and controlling it's color would be even more useful in those cases.

The problem here is that this is based on a hypothetical scenario. Sure, displaying a path could be useful some times, but the Paht2D node does not have any method to draw in-game. If you want that you'd need to roll your own function anyway and if you're making a draw function, you can make a custom color property as well.

No need to bloat the core node because of the thought that some *might someday* use it.

**Comment by willvincent** [comment]
Whatever the case, as it exists currently it's hardcoded, and that's crap. If you want to fix it some other way, by all means do. But it should _not_ be hardcoded and impossible to change.

**Pull request closed without merging**